layerindex/tools/import_layers.py

```diff
@@ -0,0 +1,494 @@
+                if appends_url:
+                    import_child_items(layerbranch,
+                                       BBAppend,
+                                       url=appends_url,
+                                       parent_orig_id=layerbranchjs['id'],
+                                       exclude_fields=['id', 'layerbranch', 'updated'],
+                                       key_fields=['filename'])
+
+                if incfiles_url:
+                    import_child_items(layerbranch,
+                                       IncFile,
+                                       url=incfiles_url,
+                                       parent_orig_id=layerbranchjs['id'],
+                                       exclude_fields=['id', 'layerbranch', 'updated'],
+                                       key_fields=['path'])
+
+            for idv in existing_layerbranches:
+                layerbranch = LayerBranch.objects.get(id=idv)
+                if layer_re is None or layer_re.match(layerbranch.layer.name):
+                    logger.debug('Deleting layerbranch %s' % layerbranch)
+                    layerbranch.delete()
+
+            # Get layer dependencies
+            logger.info('Importing layer dependencies')
+            rq = urllib.request.Request(layerdeps_url)
+            data = urllib.request.urlopen(rq).read()
+            jsdata = json.loads(data.decode('utf-8'))
+
+            exclude_fields = ['id', 'layerbranch', 'dependency', 'updated']
+            existing_deps = []
+            for layerbranch in layerbranch_idmap.values():
+                existing_deps += list(LayerDependency.objects.filter(layerbranch=layerbranch).values_list('id', flat=True))
+            for layerdepjs in jsdata:
+                layerbranch = layerbranch_idmap.get(layerdepjs['layerbranch'], None)
+                if not layerbranch:
+                    # We didn't import this layerbranch, skip it
+                    continue
+                dependency = layer_idmap.get(layerdepjs['dependency'], None)
+                if not dependency:
+                    # We didn't import the dependency, skip it
+                    continue
+
+                layerdep, created = LayerDependency.objects.get_or_create(layerbranch=layerbranch, dependency=dependency)
+                if not created and layerdep.id in existing_deps:
+                    existing_deps.remove(layerdep.id)
+                for key, value in layerdepjs.items():
+                    if key in exclude_fields:
+                        continue
+                    setattr(layerdep, key, value)
+                layerdep.save()
+            for idv in existing_deps:
+                LayerDependency.objects.filter(id=idv).delete()
+
+            def import_items(desc, url, exclude_fields, objclass, idmap, parentfield):
+                logger.debug('Importing %s' % desc)
+                rq = urllib.request.Request(url)
+                data = urllib.request.urlopen(rq).read()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *urlopen*. The resource is allocated by call *request.urlopen*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close the object returned by *urlopen()* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).